controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -205,109 +193,167 @@ public void refresh() {
     @Override
     public CompletableFuture<Void> checkScopeExists() {
         return store.checkExists(scopePath)
-                .thenAccept(x -> {
-                    if (!x) {
-                        throw StoreException.create(StoreException.Type.DATA_NOT_FOUND, scopePath);
-                    }
-                });
+                    .thenAccept(x -> {
+                        if (!x) {
+                            throw StoreException.create(StoreException.Type.DATA_NOT_FOUND, scopePath);
+                        }
+                    });
     }
 
     @Override
-    CompletableFuture<Void> createSealedSegmentsRecord(byte[] sealedSegmentsRecord) {
-        return Futures.toVoid(store.createZNodeIfNotExist(sealedSegmentsPath, sealedSegmentsRecord));
+    CompletableFuture<Void> createRetentionSetDataIfAbsent(byte[] data) {
+        return Futures.toVoid(store.createZNodeIfNotExist(retentionSetPath, data));
     }
 
     @Override
-    CompletableFuture<Data> getSealedSegmentsRecord() {
-        return store.getData(sealedSegmentsPath);
+    CompletableFuture<Data> getRetentionSetData() {
+        return store.getData(retentionSetPath);
     }
 
     @Override
-    CompletableFuture<Void> updateSealedSegmentsRecord(Data update) {
-        return store.setData(sealedSegmentsPath, update)
-                .thenApply(x -> cache.invalidateCache(sealedSegmentsPath));
+    CompletableFuture<Version> updateRetentionSetData(Data retention) {
+        return store.setData(retentionSetPath, retention)
+                .thenApply(Version.IntVersion::new);
     }
 
     @Override
-    CompletableFuture<Void> createRetentionSet(byte[] retention) {
-        return Futures.toVoid(store.createZNodeIfNotExist(retentionPath, retention));
+    CompletableFuture<Void> createStreamCutRecordData(long recordingTime, byte[] record) {
+        String path = String.format(retentionStreamCutRecordPathFormat, recordingTime);
+        return Futures.toVoid(store.createZNodeIfNotExist(path, record));
     }
 
     @Override
-    CompletableFuture<Data> getRetentionSet() {
-        return store.getData(retentionPath);
+    CompletableFuture<Data> getStreamCutRecordData(long recordingTime) {
+        String path = String.format(retentionStreamCutRecordPathFormat, recordingTime);
+        return cache.getCachedData(path);
     }
 
     @Override
-    CompletableFuture<Void> updateRetentionSet(Data retention) {
-        return store.setData(retentionPath, retention)
-                .thenApply(x -> cache.invalidateCache(retentionPath));
+    CompletableFuture<Void> deleteStreamCutRecordData(long recordingTime) {
+        String path = String.format(retentionStreamCutRecordPathFormat, recordingTime);
+
+        return store.deletePath(path, false)
```

Comment:
Should we invalidate the cache entry and then delete? In this current order, we have that possibility that the data item is a cache hit after it has been deleted from the store.

Reply:
the same race occurs in the other direction too. 
Someone could have loaded the cache after we invalidate cache but before it is deleted in the underlying store.

Reply:
So the current logic of cleaning up the cache post delete in the store at least guarantees that whenever cache invalidation happens, all requests post that will not be able to get the record of interest.. 
and anyone getting it from cache after the delete should be handled in business logic.. 
anywhere where strict correctness guarantee is required, we should ignore cached value and get latest value from the store.